rp2040-hal/src/pwm.rs

```diff
@@ -1,260 +1,324 @@
-use embedded_hal::PwmPin;
-
-macro_rules! pwm {
-    ($PWMX:ident, $pwmx:ident, [
-    $($PXi:ident: ($pxi:ident, $pwms:expr, $pins:expr, $i:expr),)+]) => {
-        $(
-
-#[doc = "Struct for any of the "]
-#[doc = $pwms]
-#[doc = " pins"]
-pub struct $PXi {
-    pin: usize
+
+/// A PWM channel (i.e. slice output) is represented here.
+pub enum PwmOutput {
+    /// The A channel of the slice
+    A,
+    /// The B channel of the slice
+    B,
 }
 
-impl $PXi {
-    #[doc = "Constructor for a PWM pin struct"]
-    pub fn new (pin: usize) -> Self {
-        let mut pin_num : usize = 255;
-        for i in 0..$pins.len() {
-            if (pin == $pins[i]) {
-                pin_num = pin;
-            }
-        }
+/// PWM clock divider modes are represented here.
+pub enum PwmClkdivMode {
+    /// Free-running counting at rate dictated by fractional divider.
+    FreeRunning,
+    /// Fractional divider is gated by the PWM B pin.
+    BHigh,
+    /// Fractional divider advances with each rising edge of the PWM B pin.
+    BRising,
+    /// Fractional divider advances with each falling edge of the PWM B pin.
+    BFalling,
+}
 
-        if (pin_num == 255) {
-            pin_num = $pins[0];
-        }
+/// The `Pwm` type respresents an abstraction of the PWM hardware block.
+pub struct Pwm {
+    regs: pac::PWM,
+}
 
-        Self {
-            pin: pin_num,
-        }
+impl Pwm {
+    /// Generate a new `Pwm` instance which takes ownersip of the provided `rp2040_pac::PWM`
+    /// object.
+    pub fn new(pac_pwm: pac::PWM, resets: &mut pac::RESETS) -> Self {
+        pac_pwm.reset_bring_up(resets);
+        Pwm { regs: pac_pwm }
     }
 
-    // TODO: This function contains all the methods that required the PWM to have an instance of PADS_BANK0, RESETS, or IO_BANK0.
-    // Since the GPIO pins take ownership of these, after the GPIO refactor, this method should be moved into gpio.rs, and the PWM
-    // will instead receive a single gpio pin.
-    fn init_io(&self, pwm: &mut pac::$PWMX, pad : &mut pac::PADS_BANK0, io : &mut pac::IO_BANK0, resets: &mut pac::RESETS) -> () {
-        //TODO: Merge these into gpio.rs split function after GPIO refactor. At the moment, this is here because these need to be reset for
-        // the PWM to work. However, because they're here, they'll be reset every time a new PWM pin is created (BAD).
-        pwm.reset_bring_up(resets);
-        io.reset_bring_up(resets);
-
-        pad.gpio[self.pin].write(|w| w.ie().set_bit());
-        pad.gpio[self.pin].write(|w| w.od().clear_bit());
-        unsafe {
-            io.gpio[self.pin].gpio_ctrl.write_with_zero(|w| w.funcsel().pwm_a_0());
+    /// Generate a `PwmSlice` instance if the provided slice index is valid. Note that the index is
+    /// 0 based.
+    pub fn slice(&self, idx: usize) -> Option<PwmSlice> {
+        if idx > 7 {
+            // The rp2040 PWM block has only 8 slices indexed from 0 to 7
+            return None;
```

Comment:
It is more idiomatic in Rust to return values from each clause of a condition than to use a guard clause like this.
I've flipped the positive condition first, though this is entirely a style choice so keeping as idx > 7 
```rust
pub fn slice(&self, idx: usize) -> Option<PwmSlice> {
    // The rp2040 PWM block has only 8 slices indexed from 0 to 7
    if idx > 7  {
        None
    } else {
        Some(PwmSlice::new(self, idx))
    }
}
```